### PR TITLE
✨ More agent values

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
## Summary
This PR exposes more agent command line flags as `agent-` flags on the controller command line.

## Related issue(s)

This addresses part of https://github.com/kubestellar/kubestellar/issues/2159

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
There is this: https://github.com/kubestellar/ocm-status-addon/issues/57#issuecomment-2128612880

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -73,4 +73,11 @@ spec:
...
         - --v={{.Values.controller.verbosity}}
-        - --agent-v={{.Values.agent.verbosity}}
+        - "{{if .Values.agent.hub_burst}}--agent-hub-burst={{.Values.agent.hub_burst}}{{end}}"
+        - "{{if .Values.agent.hub_qps}}--agent-hub-qps={{.Values.agent.hub_qps}}{{end}}"
+        - "{{if .Values.agent.local_burst}}--agent-local-burst={{.Values.agent.local_burst}}{{end}}"
+        - "{{if .Values.agent.local_qps}}--agent-local-qps={{.Values.agent.local_qps}}{{end}}"
+        - "{{if .Values.agent.log_flush_frequency}}--agent-log-flush-frequency={{.Values.agent.log_flush_frequency}}{{end}}"
+        - "{{if .Values.agent.logging_format}}--agent-logging-format={{.Values.agent.logging_format}}{{end}}"
+        - "{{if .Values.agent.v}}--agent-v={{.Values.agent.v}}{{end}}"
+        - "{{if .Values.agent.vmodule}}--agent-vmodule={{.Values.agent.vmodule}}{{end}}"
```
Comment (MikeSpreitzer) [inline config/manager/manager.yaml:83]: I am not entirely happy to duplicate this much specific knowledge here and in values.yaml. The good news is that I expect that changes in the logging flags will be rare (based on historical experience). There is another approach possible: https://github.com/kubestellar/ocm-status-addon/issues/57#issuecomment-2128612880

Comment (MikeSpreitzer) [inline config/manager/manager.yaml:83]: I am not entirely happy to duplicate this much specific knowledge here and in values.yaml. The good news is that I expect that changes in the logging flags will be rare (based on historical experience). There is another approach possible: https://github.com/kubestellar/ocm-status-addon/issues/57#issuecomment-2128612880

File: config/manager/manager.yaml [hunk 2]
```diff
@@ -73,4 +73,11 @@ spec:
         args:
         - "controller"
         - --v={{.Values.controller.verbosity}}
-        - --agent-v={{.Values.agent.verbosity}}
+        - "{{if .Values.agent.hub_burst}}--agent-hub-burst={{.Values.agent.hub_burst}}{{end}}"
```
Comment (MikeSpreitzer) [inline config/manager/manager.yaml:76]: (A) Remember that kustomize is applied to _create_ the chart template; the kustomize input has to parse as valid YAML for objects following the Kubernetes Resource Model. In other words, Go template expansion _follows_ some kube object processing, so each Go template has to be inside a string in a Kube object.

(B) I was a little surprised, but it "works" to pass empty string arguments. This does not cause any consternation. In the E2E test in CI you can see the args list for the controller in its Deployment object:

```
        - args:
          - controller
          - --v=2
          - --agent-hub-burst=7
          - ""
          - ""
          - ""
          - ""
          - ""
          - --agent-v=2
          - ""
```

You can also see the log from the controller; it has no complaints about empty strings on the command line.

You can also see the args for the agent in its Deployment in the WEC:

```
    - args:
      - agent
      - --hub-kubeconfig=/var/run/hub/kubeconfig
      - --cluster-name=cluster1
      - --addon-namespace=open-cluster-management-agent-addon
      - --hub-burst=7
      - --v=2
```

File: config/manager/manager.yaml [hunk 3]
```diff
@@ -73,4 +73,11 @@ spec:
         args:
         - "controller"
         - --v={{.Values.controller.verbosity}}
-        - --agent-v={{.Values.agent.verbosity}}
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply config/manager/manager.yaml:76]: I do not follow. Where do you see the agent prefix missing?

**Comment by francostellari** [comment]
@MikeSpreitzer personally I don't like your approach about the agent values.

The values should be defined in the `values.yaml` with their default value... then add a comment to explain them

Relying on a default value in the controller code is not very good.

This also should the use of all the if closures in the controller yaml
To enforce type and default you can do something like
`"--some-arg={{.Values.same-value-that-may-not-exist  | default 2 | int}}"`

Furthermore, shouldn't all the changes in the chart yaml be done in the https://github.com/kubestellar/ocm-status-addon/blob/main/config/manager/manager.yaml otherwise they will be overwritten by `make chart`?

**Comment by MikeSpreitzer** [comment]
@francostellari, I thought it would be nice to enable this possibility: the user of this chart does not specify some agent flags, and thus the agent Deployment object does not specify some agent flags, and thus the defaults in the agent apply --- which means that if the agent image were changed to one that has better default values, this would not be defeated. I was trying to get the point where the defaults are written down in only one place, not maintained in multiple places, also for the sake of ease of maintenance.

I see the countervailing concern for early checking. Is there some other reason that you do not like leaving the defaulting up to the agent?

What does the `int` function do if its input is unacceptable? Does the template expansion fail, or does the function produce some default int value?

**Comment by francostellari** [comment]
> @francostellari, I thought it would be nice to enable this possibility: the user of this chart does not specify some agent flags, and thus the agent Deployment object does not specify some agent flags, and thus the defaults in the agent apply --- which means that if the agent image were changed to one that has better default values, this would not be defeated. I was trying to get the point where the defaults are written down in only one place, not maintained in multiple places, also for the sake of ease of maintenance.
> 
> I see the countervailing concern for early checking. Is there some other reason that you do not like leaving the defaulting up to the agent?
> 

Since you are exposing the values to the chart, the default should be forced in the chart and not left to the agent.
It's ok to have the default just in the `values.yaml` and not in the template

> What does the `int` function do if its input is unacceptable? Does the template expansion fail, or does the function produce some default int value?

It does not say: https://helm.sh/docs/chart_template_guide/function_list/#type-conversion-functions

**Comment by MikeSpreitzer** [comment]
Regarding what `int` does on bad input: it successfully outputs 0.  See https://go.dev/play/p/eiEyPkg5Y2i for a demonstration.

File: config/manager/manager.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdettori) [reply config/manager/manager.yaml:76]: It seems to me the `--v` is for the controller (in the ITS) and `--agent-v` is for the agent, so I do not see I a problem either

**Comment by pdettori** [comment]
@MikeSpreitzer, @francostellari  is the disagreement mainly about the default values, with @francostellari arguing for having them in the chart and @MikeSpreitzer for having them in the agent ? If that is the case, it does not seem a big concern to me as long as the user can alter those settings from the chart values - is that the case @MikeSpreitzer ?

**Comment by MikeSpreitzer** [comment]
The latest rev of this PR takes the approach that @francostellari was advocating, the chart now declares the relevant values with defaults.

**Comment by MikeSpreitzer** [comment]
The force-push to c957de0 is a rebase onto `main`.

File: chart/values.yaml [hunk 4]
```diff
@@ -8,11 +8,13 @@ controller:
 
 # Command line flags for the agent
 agent:
-  hub_burst: 10 # int Allowed burst in requests/sec for accessing the hub from the agent (default 10)
```
Comment (MikeSpreitzer) [inline chart/values.yaml]: I do not know why GitHub is not highlighting this properly, but the change here is removing the redundant remark about default from the end of the line.

**Review by pdettori — APPROVED** [review]
/lgtm

**Pull request merged**